### PR TITLE
Implement loss scoring for #628

### PR DESCRIPTION
**Pull request by asberk**
This addresses issue #628 . Tests could probably be more complete. @BenjaminBossan , please let me know if/how they could be fleshed out to be more comprehensive. 

As a side note, I have limited familiarity with `PassThroughScoring`, but it seemed, in a way, like overkill. This could be because I'm not properly understanding some edge cases. Anyway, at the time, I thought it was sensible to write an accumulator class that handles the reduction of the batch-level statistics for `criterion`. In case this isn't ideal, is there an easy way to convert the current code to something that uses Skorch tools (*e.g.,* `PassThroughScoring`)?

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Thanks a lot @asberk for the PR.

I haven't taken a detailed look at your code yet but two things came to my mind:

1. The docstring of `loss_scoring` needs some clean up, I think it should stick very close to other similar functions. Also, it would be nice to document how to use that function to add a `score` method to the net, similar to the discussion. That could be added to the docstring of `loss_scoring` as well, or maybe in the FAQ section of the docs.

2. I believe from a first glance that the accumulator class is overkill. Shouldn't it be possible to capture the same functionality without a class? The `loss_scoring` could collect the losses and the batch sizes, then pass them to an accumulation function, along with the criterion. Inside that function, you could call `np.average`, similar to how `PassthroughScoring` does it.

I don't think there is a need to use `PassthroughScoring` or to try to share any code with `PassthroughScoring`, the code duplication should be minimal. What do you think about this?

Overall, this looks very good though, nice unit tests.

**Comment by asberk** [comment]
Thank you for your comments, @BenjaminBossan . 

1. I've fixed up the docstring and added an example. The example is not an exemplar as `GridSearchCV` looks for higher scores I believe. If you suggest an avenue that more simply showcases the behaviour, I'll update it to that.
2. I've removed the accumulator abstraction. There is now also a test to make sure `loss_scoring` computes the same value as `net.criterion_(y_proba, labels)`.

File: skorch/utils.py [hunk 1]
```diff
@@ -544,6 +544,7 @@ class TeeGenerator:
     generator more than once.
 
     """
+
```
Comment (BenjaminBossan) [inline skorch/utils.py:547]: Could you please undo this change? I think it was accidentally checked in.

File: skorch/scoring.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    ... )
+    >>> grid_searcher.fit(X, y)
+    >>> best_net = grid_searcher.best_estimator_
+    >>> print(best_net.score(X, y))
+
+    Parameters
+    ----------
+    net : skorch.NeuralNet
+        A fitted Skorch :class:`.NeuralNet` object.
+    X : input data, compatible with skorch.dataset.Dataset
```
Comment (BenjaminBossan) [inline skorch/scoring.py:45]: Could you please add an empty line between parameters?

File: skorch/tests/test_scoring.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import torch
+
+
+class TestLossScoring:
+    @pytest.fixture(scope="module")
+    def data(self, classifier_data):
+        return classifier_data
+
+    @pytest.fixture(scope="module")
+    def val_data(self, classifier_data):
```
Comment (BenjaminBossan) [inline skorch/tests/test_scoring.py:12]: I think for testing this function, we don't actually need separate validation data, we can just use the training data.

File: skorch/scoring.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
+import numpy as np
+from skorch.net import NeuralNet
+from skorch.dataset import unpack_data
+
+
+def loss_scoring(net: NeuralNet, X, y=None):
+    """
```
Comment (BenjaminBossan) [inline skorch/scoring.py:7]:
```suggestion
    """Calculate score using the criterion of the net
    
    Use the exact same logic as during model training to calculate the score.
```

File: skorch/scoring.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    loss_scoring(net, X, y=None)
+
+    Computes the loss of ``net`` on data (``X``, ``y``).
```
Comment (BenjaminBossan) [inline skorch/scoring.py:10]:
```suggestion
```

File: skorch/scoring.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+    This function can be used to implement the ``score`` method for a
+    :class:`.NeuralNet` through sub-classing. This is useful, for example, when
+    combining skorch models with sklearn objects that rely on the model's
+    ``score`` method. Below is an example using GridSearchCV.
+
+    >>> class ScoredNet(skorch.NeuralNetClassifier):
```
Comment (BenjaminBossan) [inline skorch/scoring.py:17]: Thanks for adding the elaborate example. I think, however, that it could be simplified. 

1) You could just show how to subclass the net here.
2) You could go with the grid search example, but in that case you actually don't need to subclass the net. You can pass the `scoring=loss_scoring` to `GridSearchCV` and it should result in the same outcome.

Assume that you go with the second option: I would then add the subclassing example to the docs, probably to the FAQ ("How to implement a score method on the net that returns the loss?" or something along these lines).

File: skorch/scoring.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    ...     def score(self, X, y=None, lower_is_better=False):
```
Comment (BenjaminBossan) [inline skorch/scoring.py:18]:
```suggestion
    ...     def score(self, X, y=None):
```

I would remove this because it doesn't conform to the sklearn API. Also, if `lower_is_better = True`, which is always true for PyTorch criteria, we should return `-loss_value` and not vice versa? Otherwise, something like grid search would choose the model with the _highest_ loss.

File: skorch/scoring.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+    Returns
+    -------
+    loss_value : float32 or np.ndarray
+        Return type depends on net.criterion_.reduction, and will be a float if
+        reduction is ``'sum'`` or ``'mean'``. If reduction is ``'none'`` then
+        this function returns a ``np.ndarray`` object.
+
+    """
+    net.check_is_fitted()
```
Comment (BenjaminBossan) [inline skorch/scoring.py:68]: Good attention to detail

File: skorch/scoring.py (hunk 6 above)
Comment (asberk) [reply skorch/scoring.py:16]: I wound up with a combination of 1. and 2. The docstring now shows the sub-classing. The FAQ now has an example of `loss_scoring(net, X, y)` as well as an example using `GridSearchCV`.

**Comment by asberk** [comment]
I believe I have addressed the changes you've mentioned, @BenjaminBossan . Thank you for the detailed list and comments!

Funny enough, there were tests for `reduction` ∈ {`'mean'` `'sum'`} in that class that no longer exists, and I forgot to transfer that over. The `reduction` fixture uses `'mean'` and `'sum'` only, and `test_scored_net_matches_criterion_value` tests score using the same `reduction` that was used to train the network. This could be modified if you think it is insufficient: so that a trained net's scoring is tested for both `'mean'` and `'sum'`. 

I made a separate test for when the reduction is `'none'` since the behaviour is so different, and since the reduction must be altered after fitting. There is also a test asserting that an unknown reduction raises.

File: docs/user/FAQ.rst [hunk 9]
```diff
@@ -357,3 +357,50 @@ layer, and that the number of features can be determined by
 ``X.shape[1]``. If those assumptions are not true for your case,
 adjust the code accordingly. A fully working example can be found
 on `stackoverflow <https://stackoverflow.com/a/60170023/1643939>`_.
+
+How do I implement a score method on the net that returns the loss?
+-------------------------------------------------------------------
+
+Sometimes, it is useful to be able to compute the loss of a net from within
+``skorch`` (*e.g.,* when a net is part of an ``sklearn`` pipeline). The function
```
Comment (BenjaminBossan) [inline docs/user/FAQ.rst:365]:
```suggestion
``skorch`` (e.g. when a net is part of an ``sklearn`` pipeline). The function
```
Or why should "e.g." be in italics?

File: docs/user/FAQ.rst [hunk 10]
```diff
@@ -357,3 +357,50 @@ layer, and that the number of features can be determined by
...
+    module = nn.Sequential(
+        nn.Linear(25, 25),
+        nn.ReLU(),
+        nn.Linear(25, 2),
+        nn.Softmax(dim=1)
+    )
+    net = skorch.NeuralNetClassifier(module).fit(X, y)
+    print(loss_scoring(net, X, y))
+
+The second example shows how to sub-class :class:`skorch.NeuralNetClassifier` to
```
Comment (BenjaminBossan) [inline docs/user/FAQ.rst:386]:
```suggestion
The second example shows how to sub-class :class:`skorch.classifier.NeuralNetClassifier` to
```

At least for me that was necessary to make the link work.

File: skorch/scoring.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+            "sample_weight for loss_scoring is not yet supported."
+        )
+
+    net.check_is_fitted()
+
+    dataset = net.get_dataset(X, y)
+    iterator = net.get_iterator(dataset, training=False)
+    history = {"loss": [], "batch_size": []}
+    reduction = net.criterion_.reduction
+    assert reduction in ["mean", "sum", "none",], (
```
Comment (BenjaminBossan) [inline skorch/scoring.py:63]: Could you please raise a `ValueError` instead?

File: skorch/tests/test_scoring.py [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    def test_scored_net_with_reduction_none(
+        self, scored_net_cls, module_cls, reduction, data
+    ):
+        X, y = data
+        net = scored_net_cls(
+            module_cls, lr=0.01, criterion__reduction=reduction
+        ).fit(X, y)
+        net.set_params(criterion__reduction="none")
+        output = net.score(X, y)
+        assert output.shape[0] == X.shape[0]
```
Comment (BenjaminBossan) [inline skorch/tests/test_scoring.py:106]: I believe this test can be made a little bit more sophisticated than just checking the shape. If you first calculate the loss with `reduction='sum'` and then with `'none'`, the sum of the latter should equal the former.

File: docs/user/FAQ.rst (hunk 9 above)
Comment (asberk) [reply docs/user/FAQ.rst:365]: I'd italicized it because it's Latin, not English. But it does not matter on the whole, and I admit it is better to opt for consistency. I suppose better readability of the plaintext is a bonus of no italics.

File: skorch/tests/test_scoring.py (hunk 12 above)
Comment (asberk) [reply skorch/tests/test_scoring.py:108]: Good idea. I kept the `assert` for `shape` equality, and added the extra lines for asserting value equality.

File: docs/user/FAQ.rst (hunk 9 above)
Comment (BenjaminBossan) [reply docs/user/FAQ.rst]: Okay, I didn't know that. As you said, for consistency (and since it's easier to type), let's do without.

**Comment by BenjaminBossan** [comment]
Thanks @asberk for this feature, you did excellent work. I'm sure many users will find it useful.

**Comment by asberk** [comment]
Thank you @BenjaminBossan for the learning opportunities. Happy to contribute.
